### PR TITLE
Allow threshold function to handle multi-channel image

### PR DESCRIPTION
**Pull request by tomoaki0705**
* remove the run time check of channel number of input image
* treat the input image as single channel

### CODE REVIEW

**Comment by tomoaki0705** [comment]
The threshold function of CPU handles the multi-channel image as input.
This PR will allow CUDA version of threshold to treat multi-channel image, too.

File: modules/cudaarithm/src/cuda/threshold.cu [hunk 1]
```diff
@@ -101,8 +101,9 @@ double cv::cuda::threshold(InputArray _src, OutputArray _dst, double thresh, dou
 
     const int depth = src.depth();
 
-    CV_Assert( src.channels() == 1 && depth <= CV_64F );
+    CV_Assert( depth <= CV_64F );
     CV_Assert( type <= 4 /*THRESH_TOZERO_INV*/ );
+    src = src.reshape(src.channels(), src.rows);
```
Comment (jet47) [inline modules/cudaarithm/src/cuda/threshold.cu]: The `reshape` call will leave `src` mat with the same number of channels as it was.

IMO, you need to call

```
src = src.reshape(1);
dst = dst.reshape(1);
```

after `dst` mat initialization.

Comment (tomoaki0705) [inline modules/cudaarithm/src/cuda/threshold.cu]: Right.  Thanks for the correction.  I should have checked this.

File: modules/cudaarithm/src/cuda/threshold.cu [hunk 2]
```diff
@@ -101,8 +101,9 @@ double cv::cuda::threshold(InputArray _src, OutputArray _dst, double thresh, dou
 
     const int depth = src.depth();
 
-    CV_Assert( src.channels() == 1 && depth <= CV_64F );
+    CV_Assert( depth <= CV_64F );
     CV_Assert( type <= 4 /*THRESH_TOZERO_INV*/ );
+    src = src.reshape(1);
```
Comment (jet47) [inline modules/cudaarithm/src/cuda/threshold.cu]: Please call this reshape after `dst` mat initialization. `dst` mat should have same number of channels as original input. So call `src = src.reshape(1);` after `GpuMat dst` initialization and call `reshape` for the `dst` mat too.

**Comment by jet47** [comment]
Could you extend existed tests for the threshold function with multi-channel cases?

**Comment by vpisarev** [comment]
@jet47, @tomoaki0705, shall we wait for the tests?

**Comment by jet47** [comment]
@tomoaki0705 Please fix my remark (https://github.com/Itseez/opencv/pull/3647#discussion_r23545712).

**Comment by tomoaki0705** [comment]
Vladislav, Vadim.
This is tomoaki0705.

I'm sorry but currently I'm super busy.
In a month, I'll be able to do some commits, but currently, I have no time
to fix this.

I'm sorry about this, and I agree that the test is needed.
Just it's not possible now.

Best regards.

2015-02-20 16:59 GMT+09:00 Vladislav Vinogradov <notifications@github.com>:

> @tomoaki0705 <https://github.com/tomoaki0705> Please fix my remark (#3647
> (comment)
> <https://github.com/Itseez/opencv/pull/3647#discussion_r23545712>).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Itseez/opencv/pull/3647#issuecomment-75202180>.
>



-- 
Tomoaki Teshima <tomoaki.teshima@gmail.com>

**Comment by vpisarev** [comment]
that's fine. @jet47, would we rather wait for the test or apply the patch and add the test later?

**Comment by jet47** [comment]
I think we can wait.

**Comment by vpisarev** [comment]
ok, let's close it for now

**Pull request closed without merging**